openfl-tutorials/interactive_api_tutorials_experimental/director_kvasir/envoy_folder/kvasir_shard_descriptor.py

```diff
@@ -0,0 +1,132 @@
+# Copyright (C) 2020-2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+"""Kvasir shard descriptor."""
+
+
+import os
+from hashlib import sha384
+from pathlib import Path
+
+import numpy as np
+from PIL import Image
+
+from openfl.interface.interactive_api.shard_descriptor import ShardDescriptor
+
+
+class KvasirShardDescriptor(ShardDescriptor):
+    """Shard descriptor class."""
+
+    def __init__(self, data_folder: str = 'kvasir_data',
+                 rank_worldsize: str = '1,1',
+                 enforce_image_hw: str = None) -> None:
+        """Initialize KvasirShardDescriptor."""
+        super().__init__()
+
+        self.data_folder = Path.cwd() / data_folder
+        self.download_data(self.data_folder)
+
+        # Settings for resizing data
+        self.enforce_image_hw = None
+        if enforce_image_hw is not None:
+            self.enforce_image_hw = tuple(int(size) for size in enforce_image_hw.split(','))
+        # Settings for sharding the dataset
+        self.rank_worldsize = tuple(int(num) for num in rank_worldsize.split(','))
+
+        self.images_path = self.data_folder / 'segmented-images' / 'images'
+        self.masks_path = self.data_folder / 'segmented-images' / 'masks'
+
+        self.images_names = [img_name for img_name in sorted(os.listdir(
+            self.images_path)) if len(img_name) > 3 and img_name[-3:] == 'jpg']
+        # Sharding
+        self.images_names = self.images_names[self.rank_worldsize[0] - 1::self.rank_worldsize[1]]
+
+        # Calculating data and target shapes
+        sample, target = self[0]
+        self._sample_shape = [str(dim) for dim in sample.shape]
+        self._target_shape = [str(dim) for dim in target.shape]
+
+    @staticmethod
+    def download_data(data_folder):
+        """Download data."""
+        zip_file_path = data_folder / 'kvasir.zip'
+        os.makedirs(data_folder, exist_ok=True)
+        os.system('wget -nc'
+                  " 'https://datasets.simula.no/hyper-kvasir/hyper-kvasir-segmented-images.zip'"
+                  f' -O {zip_file_path.relative_to(Path.cwd())}')
+        zip_sha384 = 'e30d18a772c6520476e55b610a4db457237f151e' \
+                     '19182849d54b49ae24699881c1e18e0961f77642be900450ef8b22e7'
+        assert sha384(open(zip_file_path, 'rb').read(
+            os.path.getsize(zip_file_path))).hexdigest() == zip_sha384
```

Comment:
We should read file by chunks (it can be very large) to calculate its hash. Like:
```python
def sha384(filename, chunksize=1024):
    m = hashlib.sha384()
    with open(filename, 'rb') as f:
        while chunk := f.read(chunksize):
            m.update(chunk)
    return m.hexdigest()
```